### PR TITLE
AIP-84 | Public list tags API

### LINKED ISSUE #42713: AIP-84 | Public list tags endpoint
In the public API, we should have a get `/dag/tags` endpoint that return a list of all tags across all DAGs. We can then use this in the UI to filter the dags list by tag #42712

Views.py code to migrate: https://github.com/apache/airflow//blob/main/airflow/www/views.py#L1035

### PR DESCRIPTION
**Pull request by jason810496**
closes: #42713

Migrate from https://github.com/apache/airflow//blob/main/airflow/www/views.py#L1035-L1038

### CODE REVIEW

**Comment by jason810496** [comment]
Hi, this PR should be marked with the `area:API` label instead of `area:UI`. The bot automatically marked it with `area:UI` because the pre-commit process auto-generates the frontend code using OpenAPI. Thanks!

**Comment by jscheffl** [comment]
Thanks for the contribution! w/o having a full review I assume you should mark the legacy API as being migrated like e.g. in https://github.com/apache/airflow/pull/42798/files#diff-df829fc289eeca0e932974a5110a8f2c21b1962e87db7c3a3b44dacf084f54a1R47 ?

**Comment by jason810496** [comment]
Hi @jscheffl, since this API is migrated from the home view in `airflow/www/views.py,` and `dag_tags` is just a small part of the home view, I'm not sure if this is the only endpoint that needs to be migrated from the home view. @bbovenzi, do you have any insights on this?

File: airflow/api_fastapi/views/public/dags.py
Comment (rawwar) [inline airflow/api_fastapi/views/public/dags.py]: Is response_model required when we are also including return type annotation? If so, I can also add to the PRs I am working on. For already merged PRs, I don't see response_model.

**Comment by jscheffl** [comment]
> Hi @jscheffl, since this API is migrated from the home view in `airflow/www/views.py,` and `dag_tags` is just a small part of the home view, I'm not sure if this is the only endpoint that needs to be migrated from the home view. @bbovenzi, do you have any insights on this?

Okay, then if this is _not_ a migration we should clarify if this really should get to be a public API. I'd think if it is used only in one place in the new UI then we should not make it a new public API. As there are other public APIs to retrieve the sme information (https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/get_dag_details)

File: airflow/api_fastapi/views/public/dags.py
Comment (bugraoz93) [reply airflow/api_fastapi/views/public/dags.py]: I think this is not needed. ResponseModels are already automatically generated in FastAPI and cascaded to openapi spec as well. 
I think rather than adding this here, we should create a pydantic model for this response and retrun according to the new model.
Similar to this [DAGCollectionResponse](https://github.com/apache/airflow/blob/b92c66d45d206c670516f192662bfcf8ad34bec8/airflow/api_fastapi/serializers/dags.py#L103) model but without `total_entries` of course

**Comment by jason810496** [comment]
Sure, that makes sense. I’ll go ahead and move this endpoint to the UI folder.

**Review by Lee-W — APPROVED** [review]

File: airflow/api_fastapi/views/public/dags.py
Comment (pierrejeambrun) [inline airflow/api_fastapi/views/public/dags.py]: I don't think we can actually return a 400. We should remove that from the doc (create_openapi_http_exception_doc)

File: airflow/api_fastapi/views/public/dags.py
Comment (pierrejeambrun) [inline airflow/api_fastapi/views/public/dags.py]: `DagTag.name` is primary key, I don't think we need the `distinct` clause

Comment (pierrejeambrun) [inline airflow/api_fastapi/views/public/dags.py]: cc: @ephraimbuddy do we have guidelines for using the newer `select` construct vs the old `session.query` in the codebase ? Can we freely use both ?

File: airflow/api_fastapi/views/public/dags.py
Comment (pierrejeambrun) [inline airflow/api_fastapi/views/public/dags.py]: Filtering is done automatically vias the `QueryTagsFilter` through the query directly, you can check implementation of other endpoints.

Doing the filtering in memory like this is suboptimal.

File: airflow/api_fastapi/views/public/dags.py
Comment (pierrejeambrun) [inline airflow/api_fastapi/views/public/dags.py]: I do not understand the `selected` field in the response. We query tags with a filter on the names, then we should get back in response a collection of tags, maybe paginated if the request says so. 

This should be more homogenous to other list endpoints that the API have, you can take a look at them to implement something similar. (get_dags for instance)

**Review by pierrejeambrun — CHANGES_REQUESTED** [review]
Thanks for the PR, a few suggestions before we can merge this.

**Comment by pierrejeambrun** [comment]
> Okay, then if this is not a migration we should clarify if this really should get to be a public API. I'd think if it is used only in one place in the new UI then we should not make it a new public API. As there are other public APIs to retrieve the sme information (https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/get_dag_details)

I think this is a good addition to the public API because currently there is no way for a user to be able to retrieve all the `tags` available for filtering for instance. This was done in the legacy UI through custom views.

Meaning that users of the API would have to know in advance tags to effectively filter on them.

File: airflow/api_fastapi/views/public/dags.py
Comment (jason810496) [reply airflow/api_fastapi/views/public/dags.py]: According to the issue https://github.com/apache/airflow/issues/42713, the response schema should match the format defined here: https://github.com/apache/airflow/blob/main/airflow/www/views.py#L1035. The `selected` field indicates whether the tag is selected based on the query parameters in all `dag_tags`.
I think this response schema might be more convenient for the frontend ?

Comment (jason810496) [reply airflow/api_fastapi/views/public/dags.py]: However, I believe this can be handled on the frontend (by storing the user’s selected state there, instead of passing tags through the API to get the `selected` fields). It’s not necessary to migrate the `selected` logic from the legacy API.

**Comment by jason810496** [comment]
Based on all the above input, I believe the original intention of issue https://github.com/apache/airflow/issues/42713 is to provide an endpoint to retrieve all tags across all DAGs for advanced filtering functionalities. I agree with the last comment by @pierrejeambrun, so this endpoint should remain in the `public` folder. However, I think the response schema should be:

```json
{
  "tags": ["tag_1", "tag_2"],
  "total_entries": 2
}
```

Additionally, the endpoint should support `offset` and `limit` query parameters, as mentioned by @rawwar in the issue. This way, the `get_dag_tags` endpoint can use the same query parameters as other public endpoints and utilize [`paginated_select`](https://github.com/apache/airflow/blob/b92c66d45d206c670516f192662bfcf8ad34bec8/airflow/api_fastapi/views/public/dags.py#L78) on the backend.

Please correct me if I have misunderstood anything. Thanks !

**Comment by pierrejeambrun** [comment]
I think the response could be:

```
{
  "tags": [{"tag_name: "tag_1", "dag_id": "some_dag_id"}, {"tag_name": "tag_2", "dag_id": "some_dag_id"}],
  "total_entries": 2
}
```

File: airflow/api_fastapi/views/public/dags.py
Comment (bbovenzi) [reply airflow/api_fastapi/views/public/dags.py]: I think this is an area where we need to adjust the implementation a bit. Before with FAB a lot of frontend logic was done on the webserver. Now, we should just send a list of tags to the UI and the UI handles all its own selection logic.

So let's drop `selected`. But in addition to pagination, I think we'll need a `tag_name_pattern` filter param.

Sorry for the confusion!

**Comment by bbovenzi** [comment]
> I think the response could be:
> 
> ```
> {
>   "tags": [{"tag_name: "tag_1", "dag_id": "some_dag_id"}, {"tag_name": "tag_2", "dag_id": "some_dag_id"}],
>   "total_entries": 2
> }
> ```
> 
> Basically we return serialized `DagTag` in the response field.

Is `dag_id` actually valuable information though? The whole point is that multiple dags share tags, so only showing a single dag_id doesn't help anyone. I actually prefer the list of tag_name strings.

**Comment by pierrejeambrun** [comment]
I’ll check the db modelling tomorrow, something I might have missesed. I suggested that dag_id was there because that’s how the sqlalchemy model is defined.

But indeed from a business point of view this does not make sense. 

But on the other hand the name is the primary key and unique, so I guess there is something I need to double check.

**Comment by jason810496** [comment]
### Update:

The endpoint is placed in the `public` folder and supports the following query parameters:
- `tag_name_pattern`: `_SearchParam`
- `order_by`: `_OrderByParam` 
- `limit`: `QueryLimit` 
- `offset`: `QueryOffset`

Example of the response schema:
```json
{
    "tags": ["tag_1", "tag_2"],
    "total_entries": 3
}
```

For the SQL statement, the `name` field should still include `DISTINCT` since there may be rows like:
```
same_tag_name, dag_id_1
same_tag_name, dag_id_2
```
While the `name` field with the `dag_id` field creates a unique pair, the `name` field itself may have duplicates in the query results.

### Feature: `_OrderByParam` - New query parameter for ordering by a field.
- `order_by`: `asc`, `desc`
- Ordering by a single field (This could be useful for models with only a few fields, where typically sorting by a single field is sufficient.)

**Comment by kaxil** [comment]
You will have to rebase and fix conflicts as we change how `fastapi_api` folder is organized here: https://github.com/apache/airflow/pull/43062

File: airflow/api_fastapi/parameters.py
Comment (pierrejeambrun) [inline airflow/api_fastapi/parameters.py]: What is the difference with `SortParam` of that same file ?

File: airflow/api_fastapi/routes/public/dags.py
Comment (pierrejeambrun) [inline airflow/api_fastapi/routes/public/dags.py]: Yes, this is much more consistent with the rest of the endpoints.

File: airflow/api_fastapi/parameters.py
Comment (jason810496) [reply airflow/api_fastapi/parameters.py]: Since we only need the name field in DagTag, I think passing asc or desc as a query parameter for ordering is more straightforward. This could also be useful for similar use cases where only a single field needs to be ordered in a model.

Comment (pierrejeambrun) [reply airflow/api_fastapi/parameters.py]: I don't think we need that, this is really niche because most of the list endpoints have multiple order possibility. And that's some code duplication, and it is missing a few things (secondary ordering to enforce non flakyness of returned order in case the first criteria is equal)

Then as a user you have to think:
- Is there multiple sort possiblities, then use `?order_by=xxx` put a field with `-` and ` `
- There is not, then user `?order_by=xxx, with `asc` and `desc`

Mixing two concept in the same query_param.

I think we should remove it and use the systematic `order_by=field` or `order_by=-field` and I should be ready for to merge. (or at least very close to)

File: airflow/api_fastapi/routes/public/dags.py
Comment (jason810496) [reply airflow/api_fastapi/routes/public/dags.py]: Hi @pierrejeambrun, I encountered an issue with `SortParm` when working on PostgreSQL. The error details are as follows:

```
2024-10-17 16:30:52.868 UTC [15780] ERROR:  for SELECT DISTINCT, ORDER BY expressions must appear in select list at character 84
2024-10-17 16:30:52.868 UTC [15780] STATEMENT:  SELECT DISTINCT dag_tag.name 
        FROM dag_tag 
        WHERE dag_tag.name ILIKE '%%' ORDER BY CASE WHEN (dag_tag.name IS NOT NULL) THEN 0 ELSE 1 END, dag_tag.name ASC, dag_tag.name 
         LIMIT 2 OFFSET 1
```

I haven't yet thought of a solution that avoids distinguishing the dialect type for returning a specific SQL statement.

File: airflow/api_fastapi/parameters.py
Comment (jason810496) [reply airflow/api_fastapi/parameters.py]: Hi @pierrejeambrun, I encountered an issue with `SortParm` when working on PostgreSQL. The error details are as follows:

```
2024-10-17 16:30:52.868 UTC [15780] ERROR:  for SELECT DISTINCT, ORDER BY expressions must appear in select list at character 84
2024-10-17 16:30:52.868 UTC [15780] STATEMENT:  SELECT DISTINCT dag_tag.name 
        FROM dag_tag 
        WHERE dag_tag.name ILIKE '%%' ORDER BY CASE WHEN (dag_tag.name IS NOT NULL) THEN 0 ELSE 1 END, dag_tag.name ASC, dag_tag.name 
         LIMIT 2 OFFSET 1
```

I haven't yet thought of a solution that avoids distinguishing the dialect type for returning a specific SQL statement.

Comment (pierrejeambrun) [reply airflow/api_fastapi/parameters.py]: I cannot reproduce the error on your branch. Both unit tests (with sqllite), and manual local tests with breeze on postgres work:
Screenshot 2024-10-18 at 14 13 53

What are the steps to experience your issue ?

**Review by pierrejeambrun — DISMISSED** [review]
Looking good

File: airflow/api_fastapi/parameters.py
Comment (pierrejeambrun) [reply airflow/api_fastapi/parameters.py]: Yes, I think it's because of the `distinct` requirering the `case` statement in the select. `group_by` seems to solve the problem.

`select(DagTag.name).group_by(DagTag.name)` instead of `select(distinct(DagTag.name))`.

**Comment by jason810496** [comment]
### Update:
Hi @pierrejeambrun, the `group_by` resolved the issue. Thanks!

### Refactor: `SortParm`
- Moved `attr_mapping` to the `__init__` parameter (since `SortParm` is used by different endpoints).
- Consolidated `allow_attr` with `attr_mapping` (we can use `attr_mapping` to validate the query parameters).
- Added `default_attr` (for models with multiple primary keys, like `DagTag`, the original `get_primary_key_column` might not be the column we want to sort by. With `default_attr`, we can specify the default field for sorting).

**Comment by pierrejeambrun** [comment]
Thanks @jason810496, Do you mind putting the refactoring part into its own PR so it does not block this one.

File: airflow/api_fastapi/core_api/routes/public/dags.py
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/dags.py]: Most of this is not necessay. By default if the attribute is not in the mapping it will take it from the model. No need to duplicate it.

**Comment by jason810496** [comment]
> Thanks @jason810496, Do you mind putting the refactoring part into its own PR so it does not block this one.


Sure, so I should create a new branch on top of the current commit and rebase the current branch.
After the current PR is merged, I will create a PR for refactoring `SortParm` from the new branch.
Is that correct?

File: airflow/api_fastapi/core_api/routes/public/dags.py
Comment (jason810496) [reply airflow/api_fastapi/core_api/routes/public/dags.py]: Agree, I will put `allowed_attrs` back and remove the duplicate cases for `attr_mapping` in the `SortParm` refactoring PR.

**Comment by pierrejeambrun** [comment]
Yes, you can checkout another branch from here (full with refactoring). Then in this branch you can drop them (the refactoring part)

**Comment by jason810496** [comment]
Resolved! I have removed the refactoring part.

**Review by bbovenzi — APPROVED** [review]

**Review by pierrejeambrun — APPROVED** [review]

**Comment by pierrejeambrun** [comment]
Unrelated CI failure, mergind.

Thanks @jason810496

**Pull request merged**
